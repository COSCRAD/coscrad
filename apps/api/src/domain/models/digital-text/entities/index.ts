export * from './digital-text-page.entity';
export * from './digital-text.entity';
export * from './types';
