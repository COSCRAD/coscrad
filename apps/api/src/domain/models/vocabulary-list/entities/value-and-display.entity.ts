import { IValueAndDisplay } from '@coscrad/api-interfaces';
import { ApiProperty } from '@nestjs/swagger';

export class ValueAndDisplay<T> implements IValueAndDisplay<T> {
    @ApiProperty({
        example: '11',
        description: 'a string or boolean value to be used in filtering a vocabulary list',
    })
    public readonly value: T;

    @ApiProperty({
        example: 'I',
        description: 'the user-facing label for the given value',
    })
    public readonly display: string;
}
