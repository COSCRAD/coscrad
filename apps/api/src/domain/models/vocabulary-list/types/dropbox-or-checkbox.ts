export { DropboxOrCheckbox } from '@coscrad/api-interfaces';
