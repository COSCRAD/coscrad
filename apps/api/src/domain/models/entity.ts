import { PartialDTO } from 'apps/api/src/types/partial-dto';
import { EntityType } from '../types/entityType';
import { isNullOrUndefined } from '../utilities/validation/is-null-or-undefined';
import { EntityId } from './types/entity-id';
import { EntityCompositeIdentifier } from './types/entityCompositeIdentifier';

// TODO Extract to a separate location
const isValidStringWithLength = (input: unknown): input is string =>
  typeof input === 'string' && input.length > 0;

/**
 *
 * TODO move this near the corresponding type. Note that this extra layer is
 * meant to safe guard us in case we want stricter validation on an `entity id`.
 * Should we change the rule, we can update this in once place, instead of replacing
 * `isValidStringWithLength` everywhere.
 */
const isValidEntityId = (input: unknown): input is EntityId =>
  isValidStringWithLength(input);

export abstract class Entity {
  readonly id: EntityId;

  abstract readonly type: EntityType;

  readonly published: boolean;

  constructor(dto: unknown) {
    // TODO Validate DTO
    if (!this.#validateDTO(dto))
      throw new Error(`Invalid dto for an entity: ${JSON.stringify(dto)}`);

    this.id = dto.id;

    this.published = typeof dto.published === 'boolean' ? dto.published : false;
  }

  getCompositeIdentifier = (): EntityCompositeIdentifier => ({
    type: this.type,
    id: this.id,
  });

  /**
   * TODO consider moving this to a `Serializable` mixin. There may be cases
   * where we want this behaviour on a non domain-model class without the
   * inheritance baggage.
   *  */
  toDTO<TEntity extends Entity>(this: TEntity): PartialDTO<TEntity> {
    return JSON.parse(JSON.stringify(this));
  }

  /**
   * TODO develop a general pattern \ directory structure for validation \ validators.
   * One idea: Return a symbol Valid | Errors[] from a validator instead of a
   * boolean flag.
   *
   * We may break the validators out into separate files. Then we will validate
   * a DTO before passing it to the constructor in our `instance factory` methods
   * in the mapping layer for the repositories.
   *  */
  #validateDTO = (dto: unknown): dto is PartialDTO<Entity> => {
    const { id } = dto as PartialDTO<Entity>;

    // A new entity will have id auto assigned by db
    if (!isNullOrUndefined(id) && !isValidEntityId(id)) return false;

    return true;
  };
}
