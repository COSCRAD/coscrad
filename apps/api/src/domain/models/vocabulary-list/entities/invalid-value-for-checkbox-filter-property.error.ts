import { InternalError } from '../../../../lib/errors/InternalError';
import { ValueAndDisplay } from './value-and-display.entity';

export class InvalidValueForCheckboxFilterPropertyError extends InternalError {
    constructor({ value, display }: ValueAndDisplay<unknown>, name: string) {
        const msg = `You cannot add the value: ${value} (${display}) as the value of the checkbox vocabulary list filter property ${name}.. The only values that are permitted for checkbox filters are true or false.`;

        super(msg);
    }
}
