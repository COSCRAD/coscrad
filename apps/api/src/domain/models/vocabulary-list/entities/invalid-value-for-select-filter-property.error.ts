import { InternalError } from '../../../../lib/errors/InternalError';
import { ValueAndDisplay } from './value-and-display.entity';

export class InvalidValueForSelectFilterPropertyError extends InternalError {
    constructor({ value, display: display }: ValueAndDisplay<unknown>, name: string) {
        const msg = `You cannot add the value: ${value} (${display}) as the value of the select vocabulary list filter property ${name}.. Only string values are permitted for select filters.`;

        super(msg);
    }
}
