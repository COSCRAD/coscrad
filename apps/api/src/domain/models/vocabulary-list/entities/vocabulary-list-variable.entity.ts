import { NonEmptyString } from '@coscrad/data-types';
import { ApiProperty } from '@nestjs/swagger';
import { DropboxOrCheckbox } from '../../vocabulary-list/types/dropbox-or-checkbox';
import { VocabularyListVariableValue } from '../../vocabulary-list/types/vocabulary-list-variable-value';
import { ValueAndDisplay } from './value-and-display.entity';

/**
 * TODO correlate `type` and `validValues`
 * type:dropbox <-> string values
 * type:checkbox <-> boolean values
 */
export class VocabularyListVariable<
    TVariableType extends VocabularyListVariableValue = VocabularyListVariableValue
> {
    @ApiProperty({
        example: 'person',
        description: 'name of a property that parametrizes terms in the list',
    })
    @NonEmptyString()
    name: string;

    @ApiProperty({
        example: 'dropbox',
        description:
            'specifies whether the corresponding field be a dropbox (select) or slider (switch)',
    })
    // TODO Support `DropboxOrCheckbox` enum data-type
    type: DropboxOrCheckbox;

    @ApiProperty({
        description: 'specifies the value and label for the corresponding form element',
    })
    // TODO Add data-type decorator
    validValues: ValueAndDisplay<TVariableType>[];
}
