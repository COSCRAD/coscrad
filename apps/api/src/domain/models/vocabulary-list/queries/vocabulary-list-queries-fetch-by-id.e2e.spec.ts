import {
    AggregateType,
    CoscradUserRole,
    IDetailQueryResult,
    IMultilingualTextItem,
    IVocabularyListViewModel,
    LanguageCode,
} from '@coscrad/api-interfaces';
import { INestApplication } from '@nestjs/common';
import 'reflect-metadata';
import * as request from 'supertest';
import httpStatusCodes, { HttpStatusCode } from '../../../../app/constants/httpStatusCodes';
import setUpIntegrationTest from '../../../../app/controllers/__tests__/setUpIntegrationTest';
import { ArangoDatabaseProvider } from '../../../../persistence/database/database.provider';
import TestRepositoryProvider from '../../../../persistence/repositories/__tests__/TestRepositoryProvider';
import generateDatabaseNameForTestSuite from '../../../../persistence/repositories/__tests__/generateDatabaseNameForTestSuite';
import {
    buildTestViewData,
    VocabularyListViewModel,
} from '../../../../queries/buildViewModelForResource/viewModels';
import { TermViewModel } from '../../../../queries/buildViewModelForResource/viewModels/term.view-model';
import getValidAggregateInstanceForTest from '../../../__tests__/utilities/getValidAggregateInstanceForTest';
import { buildMultilingualTextFromBilingualText } from '../../../common/build-multilingual-text-from-bilingual-text';
import { MultilingualText } from '../../../common/entities/multilingual-text';
import { AggregateId } from '../../../types/AggregateId';
import buildDummyUuid from '../../__tests__/utilities/buildDummyUuid';
import { AccessControlList } from '../../shared/access-control/access-control-list.entity';
import { CoscradUserWithGroups } from '../../user-management/user/entities/user/coscrad-user-with-groups';
import { CoscradUser } from '../../user-management/user/entities/user/coscrad-user.entity';
import {
    IVocabularyListQueryRepository,
    VOCABULARY_LIST_QUERY_REPOSITORY_TOKEN,
} from './vocabulary-list-query-repository.interface';

const vocabularyListId = buildDummyUuid(4);

const termId = buildDummyUuid(5);

const termOriginalText = 'I swim';

const termOriginalLanguageCode = LanguageCode.Chilcotin;

const termTranslation = 'I am swimming (English)';

const termTranslationLanguageCode = LanguageCode.English;

const indexEndpoint = `/resources/vocabularyLists`;

const buildDetailEndpoint = (id: AggregateId) => `${indexEndpoint}/${id}`;

/**
 * TODO We may want to event source our test setup to give us a very high
 * level of confidence that our back-end is working from event handler
 * to query endpoint. Eagerly joining in terms, tags, notes, etc. into vocabulary
 * lists, for example, is a big risk.
 */
const publishedTerm: TermViewModel = TermViewModel.fromDto({
    id: termId,
    isPublished: true,
    accessControlList: new AccessControlList().toDTO(),
    actions: [],
    name: buildMultilingualTextFromBilingualText(
        {
            text: termOriginalText,
            languageCode: termOriginalLanguageCode,
        },
        {
            text: termTranslation,
            languageCode: termTranslationLanguageCode,
        }
    ),
    contributions: [],
});

// TODO check that contributors are joined in- let's do this now
// const existingContributor = getValidAggregateInstanceForTest(AggregateType.contributor)

const publishedVocabularyList = buildTestViewData(VocabularyListViewModel, {
    id: vocabularyListId,
    isPublished: true,
    entries: [
        {
            term: publishedTerm,
            variableValues: {},
        },
    ],
});

const publishedVocabularyListWithUnpublishedTerm = publishedVocabularyList.clone({
    isPublished: true,
    entries: [
        {
            term: publishedTerm.clone({
                isPublished: false,
            }),
            variableValues: {},
        },
    ],
});

/**
 * TODO[test-coverage]: include test case with import history
 *
 * This test is being reworked on a separate branch. Opt back in when rebasing.
 */

const testUserId = buildDummyUuid(55);

const testUserGroupId = buildDummyUuid(56);

const testUser = getValidAggregateInstanceForTest(AggregateType.user).clone({
    id: testUserId,
    roles: [CoscradUserRole.viewer],
});

const testUserGroup = getValidAggregateInstanceForTest(AggregateType.userGroup).clone({
    id: testUserGroupId,
    userIds: [testUserId, buildDummyUuid(876)],
});

const nonAdminUserWithGroups = new CoscradUserWithGroups(testUser, [testUserGroup]);

const _privateVocabularyListWithUserPermissions = publishedVocabularyList.clone({
    accessControlList: new AccessControlList({
        allowedUserIds: [nonAdminUserWithGroups.id, buildDummyUuid(908)],
        allowedGroupIds: [nonAdminUserWithGroups.groups[0].id, buildDummyUuid(909)],
    }),
});

interface AssertQueryResultParams<TResponseBody = unknown> {
    app: INestApplication;
    seedInitialState: () => Promise<void>;
    endpoint: string;
    expectedStatus: HttpStatusCode;
    checkResponseBody?: (body: TResponseBody) => Promise<void>;
}

const assertQueryResult = async ({
    app,
    seedInitialState,
    endpoint,
    expectedStatus,
    checkResponseBody,
}: AssertQueryResultParams) => {
    await seedInitialState();

    const res = await request(app.getHttpServer()).get(endpoint);

    expect(res.status).toBe(expectedStatus);

    if (typeof checkResponseBody === 'function') {
        await checkResponseBody(res.body);
    }
};

describe(`when querying for a vocabulary list: fetch by ID`, () => {
    const testDatabaseName = generateDatabaseNameForTestSuite();

    let app: INestApplication;

    let testRepositoryProvider: TestRepositoryProvider;

    let databaseProvider: ArangoDatabaseProvider;

    const buildTestModule = async (userWithGroups: CoscradUserWithGroups) => {
        // TODO avoid using `setUpIntegrationTest` here
        ({ app, testRepositoryProvider, databaseProvider } = await setUpIntegrationTest(
            {
                ARANGO_DB_NAME: testDatabaseName,
            },
            {
                testUserWithGroups: userWithGroups,
            }
        ));
    };

    beforeEach(async () => {
        await testRepositoryProvider.testSetup();
    });

    afterAll(async () => {
        await app.close();

        databaseProvider.close();
    });

    describe(`when the user is unauthenticated`, () => {
        beforeAll(async () => {
            // TODO avoid using `setUpIntegrationTest` here
            ({ app, testRepositoryProvider, databaseProvider } = await setUpIntegrationTest(
                {
                    ARANGO_DB_NAME: testDatabaseName,
                }
                // no authenticated user
            ));
        });

        describe(`when there is a vocabulary list with the given ID`, () => {
            describe(`when the vocabulary list is published`, () => {
                describe(`when the term which appears as an entry is published`, () => {
                    it(`should return the correct vocabulary list view`, async () => {
                        await assertQueryResult({
                            app,
                            seedInitialState: async () => {
                                await app
                                    .get<IVocabularyListQueryRepository>(
                                        VOCABULARY_LIST_QUERY_REPOSITORY_TOKEN
                                    )
                                    .create(publishedVocabularyList);
                            },
                            endpoint: buildDetailEndpoint(vocabularyListId),
                            expectedStatus: HttpStatusCode.ok,
                            checkResponseBody: async (
                                body: IDetailQueryResult<IVocabularyListViewModel>
                            ) => {
                                const { name: nameDto, entries } = body;

                                const foundName = new MultilingualText(nameDto);

                                const { text: foundText, languageCode: foundLanguageCode } =
                                    foundName.getOriginalTextItem();

                                expect(foundText).toBe(publishedVocabularyList.name.items[0].text);

                                expect(foundLanguageCode).toBe(
                                    publishedVocabularyList.name.items[0].languageCode
                                );

                                // here we check that the term comes through
                                expect(entries).toHaveLength(1);

                                const { term: foundTerm } = entries[0];

                                const { name: termNameDto } = foundTerm;

                                const termName = new MultilingualText(termNameDto);

                                const { text: foundTermText, languageCode: foundTermLanguageCode } =
                                    termName.getOriginalTextItem();

                                expect(foundTermText).toBe(termOriginalText);

                                expect(foundTermLanguageCode).toBe(termOriginalLanguageCode);

                                const {
                                    text: foundTermTranslationText,
                                    languageCode: foundTermTranslationLanguageCode,
                                } = termName.getTranslation(
                                    termTranslationLanguageCode
                                ) as IMultilingualTextItem;

                                expect(foundTermTranslationText).toBe(termTranslation);

                                expect(foundTermTranslationLanguageCode).toBe(
                                    termTranslationLanguageCode
                                );

                                /**
                                 * We do this once and only once as a test of
                                 * the contract with the client. If this snapshot changes,
                                 * it means we have changed the contract with the client,
                                 * and the front-end may require corresponding changes.
                                 */
                                expect(body).toMatchSnapshot();
                            },
                        });
                    });
                });

                describe(`when the term which appears as an entry is **not** published`, () => {
                    it(`should return the correct vocabulary list view`, async () => {
                        await assertQueryResult({
                            app,
                            endpoint: buildDetailEndpoint(vocabularyListId),
                            expectedStatus: HttpStatusCode.ok,
                            seedInitialState: async () => {
                                await app
                                    .get<IVocabularyListQueryRepository>(
                                        VOCABULARY_LIST_QUERY_REPOSITORY_TOKEN
                                    )
                                    .create(publishedVocabularyListWithUnpublishedTerm);
                            },
                            checkResponseBody: async (
                                body: IDetailQueryResult<IVocabularyListViewModel>
                            ) => {
                                const { name: namedto, entries } = body;

                                const foundName = new MultilingualText(namedto);

                                const { text: foundText, languageCode: foundLanguageCode } =
                                    foundName.getOriginalTextItem();

                                expect(foundText).toBe(publishedVocabularyList.name.items[0].text);

                                expect(foundLanguageCode).toBe(
                                    publishedVocabularyList.name.items[0].languageCode
                                );

                                /**
                                 * Note that the following is crucial. Because the
                                 * term is not published, it should not be returned
                                 * from the query.
                                 */
                                expect(entries).toHaveLength(0);
                            },
                        });
                    });
                });
            });

            describe(`when the vocabulary list is not published`, () => {
                it(`should return not found`, async () => {
                    await assertQueryResult({
                        app,
                        seedInitialState: async () => {
                            await app
                                .get<IVocabularyListQueryRepository>(
                                    VOCABULARY_LIST_QUERY_REPOSITORY_TOKEN
                                )
                                .create(
                                    publishedVocabularyList.clone({
                                        isPublished: false,
                                    })
                                );
                        },
                        endpoint: buildDetailEndpoint(vocabularyListId),
                        expectedStatus: HttpStatusCode.notFound,
                    });
                });
            });
        });

        describe(`when there is no vocabulary list with the given ID`, () => {
            it(`should return not found`, async () => {
                const res = await request(app.getHttpServer()).get(
                    buildDetailEndpoint(buildDummyUuid(456))
                );

                expect(res.status).toBe(httpStatusCodes.notFound);
            });
        });
    });

    describe(`when the user is authenticated as an ordinary user`, () => {
        beforeAll(async () => {
            await buildTestModule(nonAdminUserWithGroups);
        });

        describe(`when the vocabulary list is published`, () => {
            describe(`when the term that the is subject of an entry is published`, () => {
                it('should return the given term', async () => {
                    await assertQueryResult({
                        app,
                        endpoint: buildDetailEndpoint(vocabularyListId),
                        expectedStatus: HttpStatusCode.ok,
                        seedInitialState: async () => {
                            await app
                                .get<IVocabularyListQueryRepository>(
                                    VOCABULARY_LIST_QUERY_REPOSITORY_TOKEN
                                )
                                .create(publishedVocabularyList);
                        },
                        checkResponseBody: async ({
                            entries,
                        }: IDetailQueryResult<IVocabularyListViewModel>) => {
                            expect(entries).toHaveLength(1);

                            expect(entries[0].term.id).toBe(publishedTerm.id);
                        },
                    });
                });
            });

            describe(`when the term that is subject of an entry is not published (and the user has no query ACL level permissions)`, () => {
                it(`should not return the given term`, async () => {
                    await assertQueryResult({
                        app,
                        endpoint: buildDetailEndpoint(vocabularyListId),
                        expectedStatus: HttpStatusCode.ok,
                        seedInitialState: async () => {
                            const publishedVocabularyListWithUnpublishedTerm =
                                publishedVocabularyList.clone({
                                    entries: [
                                        {
                                            term: publishedTerm.clone({
                                                isPublished: false,
                                                // empty
                                                accessControlList: new AccessControlList().toDTO(),
                                            }),
                                        },
                                    ],
                                });

                            await app
                                .get<IVocabularyListQueryRepository>(
                                    VOCABULARY_LIST_QUERY_REPOSITORY_TOKEN
                                )
                                .create(publishedVocabularyListWithUnpublishedTerm);
                        },
                        checkResponseBody: async ({
                            entries,
                        }: IDetailQueryResult<IVocabularyListViewModel>) => {
                            expect(entries).toHaveLength(0);
                        },
                    });
                });
            });

            describe(`when the term that is subject of an entry is not published, but the user has priviliged read access via a query ACL`, () => {
                it(`should return the term as one of the entries`, async () => {
                    const privateTermUserCanAccess = publishedTerm.clone({
                        isPublished: false,
                        accessControlList: new AccessControlList().allowUser(
                            nonAdminUserWithGroups.id
                        ),
                    });

                    await assertQueryResult({
                        app,
                        endpoint: buildDetailEndpoint(vocabularyListId),
                        expectedStatus: HttpStatusCode.ok,
                        seedInitialState: async () => {
                            await app
                                .get<IVocabularyListQueryRepository>(
                                    VOCABULARY_LIST_QUERY_REPOSITORY_TOKEN
                                )
                                .create(
                                    publishedVocabularyList.clone({
                                        isPublished: true,
                                        entries: [
                                            {
                                                term: privateTermUserCanAccess,
                                            },
                                        ],
                                    })
                                );
                        },
                        checkResponseBody: async ({
                            entries,
                        }: IDetailQueryResult<IVocabularyListViewModel>) => {
                            expect(entries).toHaveLength(1);

                            expect(entries[0].term.id).toBe(privateTermUserCanAccess.id);
                        },
                    });
                });
            });
        });

        describe(`when the vocabulary  list is not published`, () => {
            it(`should return not found`, async () => {
                await assertQueryResult({
                    app,
                    endpoint: buildDetailEndpoint(vocabularyListId),
                    expectedStatus: HttpStatusCode.notFound,
                    seedInitialState: async () => {
                        await app
                            .get<IVocabularyListQueryRepository>(
                                VOCABULARY_LIST_QUERY_REPOSITORY_TOKEN
                            )
                            .create(
                                publishedVocabularyList.clone({
                                    isPublished: false,
                                })
                            );
                    },
                });
            });
        });
    });

    describe(`when the user is a COSCRAD admin`, () => {
        beforeAll(async () => {
            await buildTestModule(
                new CoscradUserWithGroups(
                    // TODO is there a helper for this?
                    new CoscradUser({
                        id: buildDummyUuid(909),
                        authProviderUserId: `myauth|${1223}`,
                        // TODO Make it so this is not part of the DTO
                        type: AggregateType.user,
                        username: 'coscrad',
                        roles: [CoscradUserRole.superAdmin],
                    }),
                    []
                )
            );
        });

        describe(`when the vocabulary list is public`, () => {
            describe(`when all terms with entries are public`, () => {
                it(`should return the entire vocabulary list`, async () => {
                    await assertQueryResult({
                        app,
                        endpoint: buildDetailEndpoint(vocabularyListId),
                        expectedStatus: HttpStatusCode.ok,
                        seedInitialState: async () => {
                            await app
                                .get<IVocabularyListQueryRepository>(
                                    VOCABULARY_LIST_QUERY_REPOSITORY_TOKEN
                                )
                                .create(publishedVocabularyList);
                        },
                        checkResponseBody: async ({
                            entries,
                        }: IDetailQueryResult<IVocabularyListViewModel>) => {
                            expect(entries).toHaveLength(1);
                        },
                    });
                });
            });

            describe(`when one of the terms that is the subject of an entry is not public`, () => {
                it(`should return the entire vocabulary list`, async () => {
                    await assertQueryResult({
                        app,
                        endpoint: buildDetailEndpoint(vocabularyListId),
                        expectedStatus: HttpStatusCode.ok,
                        seedInitialState: async () => {
                            const unpublishedTerm = publishedTerm.clone({
                                id: buildDummyUuid(585),
                                isPublished: false,
                            });

                            const entriesWithUnpublishedTerm = [
                                {
                                    term: publishedTerm,
                                    variableValues: {},
                                },
                                {
                                    term: unpublishedTerm,
                                    variableValues: {},
                                },
                            ];

                            await app
                                .get<IVocabularyListQueryRepository>(
                                    VOCABULARY_LIST_QUERY_REPOSITORY_TOKEN
                                )
                                .create(
                                    publishedVocabularyList.clone({
                                        entries: entriesWithUnpublishedTerm,
                                    })
                                );
                        },
                        checkResponseBody: async ({
                            entries,
                        }: IDetailQueryResult<IVocabularyListViewModel>) => {
                            // the unpublished term should come through for an admin user
                            expect(entries).toHaveLength(2);
                        },
                    });
                });
            });
        });

        describe(`when the vocabulary list is not public`, () => {
            it(`should still return the vocabulary list`, async () => {
                await assertQueryResult({
                    app,
                    endpoint: buildDetailEndpoint(vocabularyListId),
                    expectedStatus: HttpStatusCode.ok,
                    seedInitialState: async () => {
                        await app
                            .get<IVocabularyListQueryRepository>(
                                VOCABULARY_LIST_QUERY_REPOSITORY_TOKEN
                            )
                            .create(
                                publishedVocabularyList.clone({
                                    isPublished: false,
                                    accessControlList: new AccessControlList().toDTO(),
                                    entries: [
                                        {
                                            term: publishedTerm.clone({
                                                isPublished: false,
                                                accessControlList: new AccessControlList().toDTO(),
                                            }),
                                            variableValues: {},
                                        },
                                    ],
                                })
                            );
                    },
                    checkResponseBody: async ({
                        entries,
                    }: IDetailQueryResult<IVocabularyListViewModel>) => {
                        expect(entries).toHaveLength(1);
                    },
                });
            });
        });
    });

    describe(`when the user is a project admin`, () => {
        beforeAll(async () => {
            // TODO avoid using `setUpIntegrationTest` here
            ({ app, testRepositoryProvider, databaseProvider } = await setUpIntegrationTest(
                {
                    ARANGO_DB_NAME: testDatabaseName,
                },
                {
                    testUserWithGroups: new CoscradUserWithGroups(
                        // TODO is there a helper for this?
                        new CoscradUser({
                            id: buildDummyUuid(909),
                            authProviderUserId: `myauth|${1223}`,
                            // TODO Make it so this is not part of the DTO
                            type: AggregateType.user,
                            username: 'coscrad',
                            roles: [CoscradUserRole.projectAdmin],
                        }),
                        []
                    ),
                }
            ));
        });

        describe(`when the vocabulary list is public`, () => {
            describe(`when all terms with entries are public`, () => {
                it(`should return the entire vocabulary list`, async () => {
                    await assertQueryResult({
                        app,
                        endpoint: buildDetailEndpoint(vocabularyListId),
                        expectedStatus: HttpStatusCode.ok,
                        seedInitialState: async () => {
                            await app
                                .get<IVocabularyListQueryRepository>(
                                    VOCABULARY_LIST_QUERY_REPOSITORY_TOKEN
                                )
                                .create(publishedVocabularyList);
                        },
                        checkResponseBody: async ({
                            entries,
                        }: IDetailQueryResult<IVocabularyListViewModel>) => {
                            expect(entries).toHaveLength(1);
                        },
                    });
                });
            });

            describe(`when one of the terms that is the subject of an entry is not public`, () => {
                it(`should return the entire vocabulary list`, async () => {
                    await assertQueryResult({
                        app,
                        endpoint: buildDetailEndpoint(vocabularyListId),
                        expectedStatus: HttpStatusCode.ok,
                        seedInitialState: async () => {
                            const unpublishedTerm = publishedTerm.clone({
                                id: buildDummyUuid(585),
                                isPublished: false,
                            });

                            const entriesWithUnpublishedTerm = [
                                {
                                    term: publishedTerm,
                                    variableValues: {},
                                },
                                {
                                    term: unpublishedTerm,
                                    variableValues: {},
                                },
                            ];

                            await app
                                .get<IVocabularyListQueryRepository>(
                                    VOCABULARY_LIST_QUERY_REPOSITORY_TOKEN
                                )
                                .create(
                                    publishedVocabularyList.clone({
                                        entries: entriesWithUnpublishedTerm,
                                    })
                                );
                        },
                        checkResponseBody: async ({
                            entries,
                        }: IDetailQueryResult<IVocabularyListViewModel>) => {
                            // the unpublished term should come through for an admin user
                            expect(entries).toHaveLength(2);
                        },
                    });
                });
            });
        });

        describe(`when the vocabulary list is not public`, () => {
            it(`should still return the vocabulary list`, async () => {
                await assertQueryResult({
                    app,
                    endpoint: buildDetailEndpoint(vocabularyListId),
                    expectedStatus: HttpStatusCode.ok,
                    seedInitialState: async () => {
                        await app
                            .get<IVocabularyListQueryRepository>(
                                VOCABULARY_LIST_QUERY_REPOSITORY_TOKEN
                            )
                            .create(
                                publishedVocabularyList.clone({
                                    isPublished: false,
                                    accessControlList: new AccessControlList().toDTO(),
                                    entries: [
                                        {
                                            term: publishedTerm.clone({
                                                isPublished: false,
                                                accessControlList: new AccessControlList().toDTO(),
                                            }),
                                            variableValues: {},
                                        },
                                    ],
                                })
                            );
                    },
                    checkResponseBody: async ({
                        entries,
                    }: IDetailQueryResult<IVocabularyListViewModel>) => {
                        expect(entries).toHaveLength(1);
                    },
                });
            });
        });
    });
});
