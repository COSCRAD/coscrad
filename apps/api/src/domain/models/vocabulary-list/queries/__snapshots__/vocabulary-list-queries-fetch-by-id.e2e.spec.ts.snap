// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when querying for a vocabulary list: fetch by ID when the user is a project admin when the vocabulary list is public when all terms with entries are public should return the entire vocabulary list 1`] = `
{
  "actions": [
    {
      "description": "Translate the name of a vocabulary list to another language",
      "form": {
        "fields": [
          {
            "constraints": [
              {
                "message": "Must be a valid \${propertyLabel}",
                "name": "IS_ENUM",
              },
              {
                "message": "Required",
                "name": "defined value",
              },
            ],
            "description": "the language in which you are naming the new vocabulary list name",
            "label": "language code",
            "name": "languageCode",
            "options": [
              {
                "display": "Chilcotin",
                "value": "clc",
              },
              {
                "display": "Haida",
                "value": "hai",
              },
              {
                "display": "English",
                "value": "en",
              },
              {
                "display": "French",
                "value": "fra",
              },
              {
                "display": "Chinook",
                "value": "chn",
              },
              {
                "display": "Zapotec",
                "value": "zap",
              },
              {
                "display": "Spanish",
                "value": "spa",
              },
            ],
            "type": "STATIC_SELECT",
          },
          {
            "constraints": [
              {
                "message": "must be a non-empty string",
                "name": "non-empty string",
              },
              {
                "message": "must be a defined value",
                "name": "defined value",
              },
            ],
            "description": "text for the translation of the name",
            "label": "translation",
            "name": "text",
            "type": "TEXT_FIELD",
          },
        ],
        "prepopulatedFields": {
          "aggregateCompositeIdentifier": {
            "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100004",
            "type": "vocabularyList",
          },
        },
      },
      "label": "Translate Vocabulary List Name",
      "type": "TRANSLATE_VOCABULARY_LIST_NAME",
    },
  ],
  "contributions": [],
  "entries": [
    {
      "term": {
        "accessControlList": {
          "allowedGroupIds": [],
          "allowedUserIds": [],
        },
        "actions": [],
        "contributions": [
          {
            "contributorIds": [],
            "date": {
              "day": 26,
              "month": "09",
              "year": 2022,
            },
            "statement": "Watchamacallit fabricated by: Willy Wankie",
            "type": "WHATCHAMACALLIT_FABRICATED",
          },
        ],
        "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100005",
        "isPublished": true,
        "name": {
          "items": [
            {
              "languageCode": "clc",
              "role": "original",
              "text": "I swim",
            },
            {
              "languageCode": "en",
              "role": "free translation",
              "text": "I am swimming (English)",
            },
          ],
        },
        "vocabularyLists": [],
      },
      "variableValues": {},
    },
  ],
  "form": {
    "fields": [],
  },
  "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100004",
  "isPublished": true,
  "name": {
    "items": [
      {
        "languageCode": "en",
        "role": "original",
        "text": "vocab list name (orig)",
      },
    ],
  },
}
`;
